src/services/exit-logs/verifier.ts

```diff
@@ -0,0 +1,239 @@
+      finalizationTx.input
+    )
+
+    const { data, refSlot, consensusVersion, requestsCount, dataFormat } =
+      finalizationDecoded.data as {
+        data: string
+        refSlot: ethers.BigNumber
+        consensusVersion: ethers.BigNumber
+        requestsCount: ethers.BigNumber
+        dataFormat: ethers.BigNumber
+      }
+
+    // Strip 0x
+    if (!data.includes((validatorPubkey as string).slice(2)))
+      throw new Error('Pubkey for exit was not found in finalized tx data')
+
+    const encodedData = ethers.utils.defaultAbiCoder.encode(
+      [
+        'tuple(uint256 consensusVersion, uint256 refSlot, uint256 requestsCount, uint256 dataFormat, bytes data)',
+      ],
+      [[consensusVersion, refSlot, requestsCount, dataFormat, data]]
+    )
+
+    const dataHash = ethers.utils.keccak256(encodedData)
+
+    const originTxHash = await consensusReachedTransactionHash(
+      toBlock,
+      refSlot.toString(),
+      dataHash
+    )
+
+    const originTx = await getTransaction(originTxHash)
+
+    const hashConsensusFragment = ethers.utils.Fragment.from(
+      'function submitReport(uint256 slot, bytes32 report, uint256 consensusVersion)'
+    )
+
+    const hashConsensusIface = new ethers.utils.Interface([
+      hashConsensusFragment,
+    ])
+
+    const submitReportDecoded = hashConsensusIface.decodeFunctionData(
+      hashConsensusFragment.name,
+      originTx.input
+    )
+
+    if (submitReportDecoded.report !== dataHash)
+      throw new Error(
+        'Report data hash mismatch detected between the original report and finalized event'
+      )
+
+    const expandedSig = {
+      r: originTx.r,
+      s: originTx.s,
+      v: parseInt(originTx.v),
+    }
+
+    const sig = ethers.utils.joinSignature(expandedSig)
+
+    const txData = {
+      gasLimit: ethers.BigNumber.from(originTx.gas),
+      maxFeePerGas: ethers.BigNumber.from(originTx.maxFeePerGas),
+      maxPriorityFeePerGas: ethers.BigNumber.from(
+        originTx.maxPriorityFeePerGas
+      ),
+      data: originTx.input,
+      nonce: parseInt(originTx.nonce),
+      to: originTx.to,
+      value: ethers.BigNumber.from(originTx.value),
+      type: parseInt(originTx.type),
+      chainId: parseInt(originTx.chainId),
+    }
+    const encodedTx = ethers.utils.serializeTransaction(txData) // RLP encoded tx
+    const hash = ethers.utils.keccak256(encodedTx)
+    const recoveredAddress = ethers.utils.recoverAddress(hash, sig)
+
+    // Address can be passed as checksummed or not, account for that
+    const allowlist = ORACLE_ADDRESSES_ALLOWLIST.map((address) =>
+      address.toLowerCase()
+    )
+    if (!allowlist.includes(recoveredAddress.toLowerCase())) {
+      logger.error('Transaction is not signed by a trusted Oracle', {
+        address: recoveredAddress,
+      })
+      throw new Error('Transaction is not signed by a trusted Oracle')
+    }
+  }
+
+  const lastRequestedValidatorIndex = async (operatorId: number) => {
```

Comment:
Looks like this function could be moved out of here

Reply:
This looks like a minor improvement, I suggest doing it in conjunction with TW
https://linear.app/lidofi/issue/ORC-355/validatorejector-tw-preparations